### PR TITLE
Fix compatibility with specification in regards to `client_mac`

### PR DESCRIPTION
**Pull request by falko17**
# Summary
This pull request fixes two bugs which lead to an incorrect `client_mac` being calculated: the first bug was about reusing a state parameter which has already been "used up", the second was using an incorrect key length.

# Details

## State parameter reuse
The function `calc_preamble` ends by calling `crypto_hash_sha512_final` on it. After this (according to the [libsodium documentation]) the passed-in state parameter *must not be re-used*. However, in our case the parameter *was* re-used, because the `preamble_state` was used not only for the `server_mac`, but also for the `client_mac`—hence, the SHA512 sum that was calculated before this commit was incorrect.
    
This bug has been fixed by copying the state parameter before passing it to `crypto_hash_sha512_final`. That way, the caller of `calc_preamble` can still use the state and the correct SHA512 sum will be calculated.
    
[libsodium documentation]: https://libsodium.gitbook.io/doc/advanced/sha-2_hash_function#notes

## Incorrect key length
The function `crypto_auth_hmacsha512` was used in `RecoverCredentials` and `CreateCredentialResponse`. This was a problem because the function (i.e., libsodium) uses 32-byte keys by default, while OPAQUE specifies a key length of 64 bytes.
    
The problem has been fixed by replacing this function call with `opaque_hmacsha512`, which uses the correct key length.

# Testing procedure / Background
I noticed these two bugs in the first place because I was using [`opaque-ke`](https://github.com/novifinancial/opaque-ke/) on one side and `libopaque` on the other. Trying to make these two libraries work together was a little difficult due to the relative complexity of pinning down why a certain step failed, but when using the version of `libopaque` from this pull request and the [latest commit of `opaque-ke`](https://github.com/novifinancial/opaque-ke/tree/5a0aef607ccda21ae8a76c687de801632dff2574)[^1], the login flow finally completes successfully.

In other words, it should now be possible to use `libopaque` and `opaque-ke` together without any problems.

(I've also run the tests generated by `make tests` to make sure I didn't break anything.)

[^1]: Before this commit (which is not yet released in the crate itself), `opaque-ke` used a non-standard salt length, which also made it impossible to combine it with `libopaque`.

### CODE REVIEW

File: src/opaque.c [hunk 1]
```diff
@@ -1304,10 +1309,10 @@ int opaque_CreateCredentialResponse(const uint8_t _pub[OPAQUE_USER_SESSION_PUBLI
...
 #endif
   if(NULL!=authU) {
-    crypto_auth_hmacsha512(authU,                               // out
-                           (uint8_t*)preamble,                  // in
-                           crypto_hash_sha512_BYTES,            // len(in)
-                           keys.km3);                           // key
+    opaque_hmacsha512(keys.km3,                       // key
+                     (uint8_t*)preamble,              // in
+                     crypto_hash_sha512_BYTES,        // len(in)
+                     authU);                          // out
```
Comment (stef) [inline src/opaque.c:1315]: woah. really? if this is true. that's very embarrassing.

Comment (stef) [reply src/opaque.c:1315]: oh, i see it's also a different function. ok. that explains the different order of params. phew.

Comment (falko17) [reply src/opaque.c:1315]: Yeah, I had to use a different function here, don't worry :D

**Comment by stef** [comment]
thank you very much for this great contribution!
